### PR TITLE
(feature) Pageview tags

### PR DESCRIPTION
**Pull request by Blaumaus**
Please describe your changes using the checklist below.

### Self-hosted support
- [x] Your feature is implemented for the selfhosted version of Swetrix
- [ ] This PR only updates the production code (e.g. paddle webhooks, CAPTCHA, blog, etc.)

### Database migrations
- [x] Clickhouse / MySQL migrations added for this PR
- [ ] No table schemas changed in this PR

### Documentation
- [x] You have updated the [documentation](https://github.com/swetrix/docs) according to your PR
- [ ] This PR did not change any publicly documented endpoints

### CODE REVIEW

File: apps/production/src/analytics/analytics.controller.ts [hunk 1]
```diff
@@ -50,6 +50,7 @@ import { PageviewsDTO } from './dto/pageviews.dto'
 import { EventsDTO } from './dto/events.dto'
 import { AnalyticsGET_DTO, ChartRenderMode } from './dto/getData.dto'
 import { GetCustomEventMetadata } from './dto/get-custom-event-meta.dto'
+import { GetPagePropertyMeta } from './dto/get-page-property-meta.dto'
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/analytics.controller.ts:53]: You forgot the prefix `...Dto`.

File: apps/production/src/analytics/analytics.controller.ts [hunk 2]
```diff
@@ -75,13 +76,7 @@ import { BotDetection } from '../common/decorators/bot-detection.decorator'
...
 import { GetCustomEventsDto } from './dto/get-custom-events.dto'
 import { GetFiltersDto } from './dto/get-filters.dto'
-import {
-  IAggregatedMetadata,
-  IFunnel,
-  IGetFunnel,
-  IUserFlow,
-  PerfMeasure,
-} from './interfaces'
+import { IFunnel, IGetFunnel, IUserFlow, PerfMeasure } from './interfaces'
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/analytics.controller.ts:79]: Do not start the interface name with `I...`

File: apps/production/src/analytics/analytics.controller.ts [hunk 3]
```diff
@@ -489,14 +485,24 @@ export class AnalyticsController {
...
     }
 
-    const customs = await this.analyticsService.processCustomEV(
-      queryCustoms,
+    const customs = await this.analyticsService.getCustomEvents(
+      filtersQuery,
       paramsData,
     )
 
+    let properties = {}
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/analytics.controller.ts:493]: Typify, please.

File: apps/production/src/analytics/analytics.controller.ts [hunk 4]
```diff
@@ -602,6 +608,31 @@ export class AnalyticsController {
     return this.analyticsService.getCustomEventMetadata(data)
   }
 
+  @Get('property')
+  @Auth([], true, true)
+  async getPagePropertyMetadata(
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/analytics.controller.ts:613]: Add Swagger decorators for the method-route.

File: apps/production/src/analytics/analytics.controller.ts [hunk 5]
```diff
@@ -602,6 +608,31 @@ export class AnalyticsController {
...
+    @Query() data: GetPagePropertyMeta,
+    @CurrentUserId() uid: string,
+    @Headers() headers: { 'x-password'?: string },
+  ): Promise<ReturnType<typeof this.analyticsService.getPagePropertyMeta>> {
+    const { pid, period } = data
+    this.analyticsService.validatePID(pid)
+
+    if (!_isEmpty(period)) {
+      this.analyticsService.validatePeriod(period)
+    }
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/analytics.controller.ts:623]: Validation must take place in the DTO.

File: apps/production/src/analytics/analytics.service.ts [hunk 6]
```diff
@@ -91,8 +93,10 @@ import {
   IOverallPerformance,
   IPageflow,
   PerfMeasure,
+  PropertiesCHResponse,
 } from './interfaces'
 import { ErrorDTO } from './dto/error.dto'
+import { GetPagePropertyMeta } from './dto/get-page-property-meta.dto'
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/analytics.service.ts:99]: You forgot the prefix `...Dto`.

File: apps/production/src/analytics/analytics.service.ts [hunk 7]
```diff
@@ -2942,6 +3043,107 @@ export class AnalyticsService {
...
+      const res = await this.getTimeBucketForAllTime(pid, period, timezone)
+
+      diff = res.diff
+      // eslint-disable-next-line prefer-destructuring
+      newTimebucket = _includes(res.timeBucket, timeBucket)
+        ? timeBucket
+        : res.timeBucket[0]
+    }
+
+    this.validateTimebucket(newTimebucket)
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/analytics.service.ts:3086]: Validation must take place in the DTO.

File: apps/production/src/analytics/dto/get-page-property-meta.dto.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { PickType } from '@nestjs/swagger'
+import { AnalyticsGET_DTO } from './getData.dto'
+
+export class GetPagePropertyMeta extends PickType(AnalyticsGET_DTO, [
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/dto/get-page-property-meta.dto.ts:4]: You forgot the prefix `...Dto`.

File: apps/production/src/analytics/dto/get-page-property-meta.dto.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  'pid',
+  'period',
+  'timeBucket',
+  'from',
+  'to',
+  'filters',
+  'timezone',
+  'filters',
+] as const) {
+  property: string
```
Comment (yevheniikulisidi) [inline apps/production/src/analytics/dto/get-page-property-meta.dto.ts:14]: Add validation and document the field.

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: apps/production/src/analytics/analytics.controller.ts (hunk 5 above)
Comment (Blaumaus) [reply apps/production/src/analytics/analytics.controller.ts:623]: To be done in a separate PR
https://github.com/Swetrix/roadmap/issues/205

File: apps/production/src/analytics/analytics.service.ts (hunk 7 above)
Comment (Blaumaus) [reply apps/production/src/analytics/analytics.service.ts:3086]: to be done in a separate PR
https://github.com/Swetrix/roadmap/issues/206

File: apps/production/src/analytics/analytics.controller.ts (hunk 2 above)
Comment (Blaumaus) [reply apps/production/src/analytics/analytics.controller.ts:79]: to be discussed later as this codestyle is widely used on the Swetrix frontend codebase and is already using in lots of other places on API.
If we are to move to interfaces without I, it should be done within a scope of another PR

**Pull request merged**